miner/proposal_builder.go

```diff
@@ -291,38 +312,66 @@ func (pb *ProposalBuilder) Register(signer *signing.EdSigner) {
 
 // Start the loop that listens to layers and build proposals.
 func (pb *ProposalBuilder) Run(ctx context.Context) error {
-	next := pb.clock.CurrentLayer().Add(1)
+	current := pb.clock.CurrentLayer()
+	next := current + 1
 	pb.logger.With().Info("started", log.Inline(&pb.cfg), log.Uint32("next", next.Uint32()))
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		case <-pb.clock.AwaitLayer(next):
-			current := pb.clock.CurrentLayer()
-			if current.Before(next) {
-				pb.logger.With().Info("time sync detected, realigning ProposalBuilder",
-					log.Uint32("current", current.Uint32()),
-					log.Uint32("next", next.Uint32()),
-				)
-				continue
-			}
-			next = current.Add(1)
-			ctx := log.WithNewSessionID(ctx)
-			if current <= types.GetEffectiveGenesis() || !pb.syncer.IsSynced(ctx) {
-				continue
-			}
-			if err := pb.build(ctx, current); err != nil {
-				if errors.Is(err, errAtxNotAvailable) {
-					pb.logger.With().
-						Debug("signer is not active in epoch", log.Context(ctx), log.Uint32("lid", current.Uint32()), log.Err(err))
-				} else {
-					pb.logger.With().Warning("failed to build proposal",
-						log.Context(ctx), log.Uint32("lid", current.Uint32()), log.Err(err),
+	var eg errgroup.Group
+	prepareDisabled := pb.cfg.activeSet.Tries == 0 || pb.cfg.activeSet.RetryInterval == 0
+	if prepareDisabled {
+		pb.logger.With().Warning("activeset will not be prepared in advance")
+	}
+	eg.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-pb.clock.AwaitLayer(next):
+				current := pb.clock.CurrentLayer()
+				if current.Before(next) {
+					pb.logger.With().Info("time sync detected, realigning ProposalBuilder",
+						log.Uint32("current", current.Uint32()),
+						log.Uint32("next", next.Uint32()),
 					)
+					continue
+				}
+				next = current.Add(1)
+				ctx := log.WithNewSessionID(ctx)
+				synced := pb.syncer.IsSynced(ctx)
+				if !prepareDisabled {
+					pb.runPrepare(ctx, current)
+					prepareDisabled = true
+				}
+				if current <= types.GetEffectiveGenesis() || !synced {
+					continue
+				}
+				if err := pb.build(ctx, current); err != nil {
+					if errors.Is(err, errAtxNotAvailable) {
+						pb.logger.With().
+							Debug("signer is not active in epoch", log.Context(ctx), log.Uint32("lid", current.Uint32()), log.Err(err))
+					} else {
+						pb.logger.With().Warning("failed to build proposal",
+							log.Context(ctx), log.Uint32("lid", current.Uint32()), log.Err(err),
+						)
+					}
 				}
 			}
 		}
+	})
+	return eg.Wait()
```

Comment:
What is the point of running it on a goroutine and block here? How about dropping `eg.Go()` and running the code synchronously like it was before?